docs/enhancements/0001-worker-pods.md

```diff
@@ -0,0 +1,223 @@
+started on a node by looking at the difference between the desired state (spec) and the current one (status).  
+It prepares a worker configuration file in a `ConfigMap` based on the actions that need to be run on the node and
+creates a Pod mounting that `ConfigMap`.  
+When the Pod has completed, the controller reads the worker's output from the same `ConfigMap` and updates the
+`ModuleNodeState` status accordingly.
+
+When updating the `ModuleNodeState` status, the controller includes a timestamp corresponding to the last time a worker
+Pod ran on the node.
+This timestamp can be used to verify if a new worker Pod needs to run when the KMM operator restarts; if it older than
+the last transition of the `Node`'s `Ready` condition, then the node may have rebooted and we potentially need to load
+the kernel modules.
+
+#### The worker Pods
+
+Worker Pods are created by the `ModuleNodeState` controller.
+They run KMM code that reads the configuration file, pulls and extracts kmod OCI images, uses `modprobe` to load kmods,
+and write some output to a file.
+Both the configuration and the output file are mounted from a `ConfigMap`.
+The `ModuleNodeState` creates the `ConfigMaps` as well as the worker Pods.
+
+Worker Pod configuration can contain the instructions to load several kmods at once.
+If one kmod fails to load, that error is written into the output file and the worker program proceeds to the next kmod.
+
+##### Security
+
+Worker Pods run a privileged container but are granted zero privilege on the API Server.
+
+##### kmod images
+
+It is worth noting that worker Pods pull images from registries **into their filesystem**, not the node's.
+That filesystem disappears as soon as the Pod is terminated.  
+This means that node image caching mechanisms cannot be leveraged, and any Pod creation will entail data transfers from
+the kmod image's registry to the Pod.
+
+##### Setting the firmware lookup path
+
+Worker Pods will be able to write the firmware lookup path to `/sys/module/firmware_class/parameters/path` each time
+they run.
+This behaviour will be controlled by an operator setting will be written as part of the worker configuration.
+
+#### High-level reconciliation loop
+
+```mermaid
+flowchart TD
+    M[Module update]
+    N[Node update]
+    NSM[NodeModuleState\nspec update]
+    P[Worker Pod starts\nrunning on Node]
+    DMO{NodeModuleState\nspec matches status?}
+    E[End]
+    
+    M --> NSM
+    N --> NSM
+    
+    NSM --> DMO
+    
+    DMO -- Yes --> E
+    DMO -- No --> P
+    
+    CR[Controller starts] --> P
+    NR[Node becomes Ready] --> P
+    
+    P -- Pod completed or failed --> RO[Controller reads\nPod output] --> OUS[Controller updates\nNodeModuleState status] --> DMO
+```
+
+#### Device plugins
+
+KMM will still run the device plugin specified in the `Module` resource as a DaemonSet; however instead of relying on
```

Comment:
why do we need a new label for Device Plugin? If we are removing dependency on module.ready label, then lets just use Module's Selector field and that's it. DevicePlugin should be able to handle the case where device file is not present

Reply:
Fair enough. Let us keep using the `kmm.node.kubernetes.io/$namespace.$module.ready` label for now.